Overlay Updates

Jmr3366 opened the pull request:
### Identify the Bug or Feature request
#5234 GMs ignore locked feature
#5235 lock can now be set/reset through `overlay`
#5239 Added better checks for INTERNAL_FRAME_PREFIX

### Possible Drawbacks
Any user initiated frame,dialog,overlay functions where INTERNAL_FRAME_PREFIX is the prefix, will now return _standard reserved word error_
### Release Notes

Examples:
```[overlay("MyOverlay", "zorder=4;"): { <html></html> }]```
Will remove `locked` if it existed.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5266)
<!-- Reviewable:end -->

Jmr3366 committed:
#5139 added prior mapID

Jmr3366 committed:
#5139 added prior mapID, fixes

Jmr3366 committed:
#5234 GMs ignore locked feature
#5235 lock can now be set/reset
#5239 frames,dialogs,overlays can no longer start with INTERNAL_FRAME_PREFIX

Jmr3366 committed:
#5234 GMs ignore locked feature
#5235 lock can now be set/reset
#5239 better check for frames,dialogs,overlays usage of INTERNAL_FRAME_PREFIX

Jmr3366 committed:
#5234 GMs ignore locked feature
#5235 lock can now be set/reset
#5239 better check for frames,dialogs,overlays usage of INTERNAL_FRAME_PREFIX

Jmr3366 committed:
#5234 GMs ignore locked feature
#5235 lock can now be set/reset
#5239 better check for frames,dialogs,overlays usage of INTERNAL_FRAME_PREFIX

Jmr3366 committed:
Spotless fixes 1

Jmr3366 committed:
Merge branch 'develop' into j_develop_oLock

Jmr3366 committed:
Final spotless
removed priorMapID code

Jmr3366 committed:
Final spotless

Jmr3366 committed:
Final spotless

Jmr3366 committed:
Final spotless

Jmr3366 committed:
Final spotless

Jmr3366 committed:
Merge branch 'develop' into j_develop_oLock

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/functions/MacroDialogFunctions.java:
```diff
@@ -75,11 +76,12 @@ public Object childEvaluate(
-    if (parameters.size() > 0 && HTMLFrameFactory.isInternalOnly(parameters.get(0).toString())) {
+    if (parameters.size() > 0
+        && (HTMLFrameFactory.isInternalOnly(parameters.get(0).toString())
+            || parameters.get(0).toString().startsWith(AppConstants.INTERNAL_FRAME_PREFIX))) {
```
This `startsWith()` condition is redundant. `HTMLFrameFactory.isInternalOnly()` already does this (more, actually) via its `substring()`/`equalsIgnoreCase()` check.

Although looking at that I'm not sure why `isInternalOnly()` uses a case-insensitive check. Maybe we should just replace the `isInternalOnly()` implementation with a simpler `startsWith()` check?

kwvanderlinde replied:
And in relation to #5239, we would actually need to restrict the meaning of `"*"` specifically in `closeOverlay()` so that it does not close internal overlays.

Jmr3366 replied:
> This `startsWith()` condition is redundant. `HTMLFrameFactory.isInternalOnly()` already does this (more, actually) via its `substring()`/`equalsIgnoreCase()` check.
> 
> Although looking at that I'm not sure why `isInternalOnly()` uses a case-insensitive check. Maybe we should just replace the `isInternalOnly()` implementation with a simpler `startsWith()` check?

I added the extra check because in my testing, I was able to create an overlay starting with the reserved word.  I can not duplicate this.  I must have had a typo in my test case.  I think it would be best to revert it back and I will remove what I added here.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/htmlframe/HTMLOverlayPanel.java:
```diff
@@ -257,12 +257,18 @@ public void showOverlay(String name, int zOrder, boolean locked, String html, Ob
+                overlayManager.setZOrder(zOrder);
+                overlays.add(overlayManager);
+              }
+              if (locked != overlayManager.getLocked()) {
```
No need for the check first. Might as well just call `setLocked()`

Jmr3366 committed:
Completed requested changes

Jmr3366 committed:
Completed requested changes

Jmr3366 committed:
Merge branch 'develop' into j_develop_oLock

kwvanderlinde commented:
@Jmr3366 Just to confirm, #5239 is out of scope for this PR now, right? If so, I'm happy with how it looks.

Jmr3366 committed:
#5239 fix

Jmr3366 commented:
@kwvanderlinde I missed putting that back in from the initial commit when I was troubleshooting things as it changes how `removeAllOverlays` functions; Let me know and I'll back it out.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/htmlframe/HTMLOverlayPanel.java:
```diff
@@ -217,25 +216,19 @@ private void removeOverlay(HTMLOverlayManager overlay) {
+      if (!overlay.getName().startsWith(AppConstants.INTERNAL_FRAME_PREFIX)) {
+        removeOverlay(overlay.getName());
+      }
+    }
```
I think this needs to stay within `Platform.runLater()`, or at least the call to `removeOverlay()` needs to be.

kwvanderlinde commented:
> @kwvanderlinde I missed putting that back in from the initial commit when I was troubleshooting things as it changes how `removeAllOverlays` functions; Let me know and I'll back it out.

No, looks good to keep it in. Just one note on threading, otherwise I'm happy with how it is.

Jmr3366 commented:
I reverted back the `removeAllOverlays`. That function is now only called in one place. I removed it from being used in the `closeOverlay("*")` use case (as I understand it, wouldn't be needed there and for me ,logically speaking, I didn't see reason to treat "*" different to a named overlay).

EDIT: Working through some things I missed . . .  :(

Jmr3366 committed:
#5239 fixes from Dev input

Jmr3366 committed:
#5239 fixes from Dev input, spotlesss

Jmr3366 committed:
#5239 fixes from Dev input, spotlesss

Jmr3366 committed:
#5239 final fixes

Jmr3366 commented:
@kwvanderlinde I _hope_ this is final . . . :)

kwvanderlinde reviewed (approved):
lgtm
